rfcs/build-system/01_convergence_devstack.md

```diff
@@ -0,0 +1,191 @@
+
+> Good enough DX:
+>
+> - rapid feedback loop
+> - one common way how to do/run things
+> - docs (reflected within tooling -> `yarn test --help` should just work)
+
+### Appetite
+
+As an initial step lets focus on resolve following (smallest possible area):
+
+1. need to build before run/test/demo
+1. unified TSConfigs
+1. unified way how to transpile javascript
+1. solution for convergence/whole repo
+
+### Solution
+
+> show don't tell / TL;DR
+> ![](https://i.imgur.com/bDwWu9v.png)
+
+#### 1. need to build before run/test/lint/demo
+
+Current state:
+
+- N\* uses wepback aliases with lerna-alias
+- OUIF uses yarn workspaces heavily (symlinks of build assets), thus after every modification outside the package that is being worked on, you need to manually rebuild to get latest changes.
+
+Convergence proposal:
+
+- define one source of truth for aliasing packages that can be leveraged consistently in all tooling (bundlers,testing,demos/storybook)
+
+**👉 tsconfig path aliases + TS solution style configs**
+
+Rather then leveraging solutions that might feel appealing (like TS Project refs etc, lerna/yarn workspace symlinks), we can leverage battle tested solution that has been in place for quite some time, thus makes its adoption compatibility with 3rd party toolign straightforward.
+
+**Pre-requirements**
+
+- typescript 3.9
+
+**Example:**
+
+> @file /tsconfig.base.json
+> "root tsconfig" (intentionally not named as tsconfig.json, so editor/IDE will pick solution config per package)
+
+```json
+{
+  "extends": "@fluentui/ts-config-common",
+  "compilerOptions": {
+    "lib": ["dom", "es2017", "dom.iterable"],
+    "rootDir": ".",
+    "baseUrl": ".",
+    "paths": {
+      "@fluentui/react-button": ["packages/react-button/src/index.ts"],
+      "@fluentui/react-hooks": ["packages/react-hooks/src/index.ts"]
+      // other aliases...
+    }
+  }
+}
+```
+
+> @file /packages/react-button/tsconfig.json
+> root package tsconfig - [solution config style](https://devblogs.microsoft.com/typescript/announcing-typescript-3-9/#solution-style-tsconfig)
+
+```json
+{
+  "extends": "../../tsconfig.base.json",
+  "compilerOptions": {
+    "types": ["node", "jest"]
+  },
+  "include": [],
+  "files": [],
+  "references": [
+    {
+      // lib type config is used for all implementation files (ambient types generation)
+      "path": "./tsconfig.lib.json"
+    },
+    {
+      // spec type config is used for tests only, thus mitigates leaking globals into implementation etc. This is not used for prod builds
+      "path": "./tsconfig.spec.json"
+    }
+  ]
+}
+```
+
+![solution TS config per environment resolution](https://i.imgur.com/U58q2Ed.png)
+
+- solution config per package enables us to author different tsconfigs per environment (unit test env -> `tsconfig.spec.json` / implementation env -> `tsconfig.lib.json`), thus properly scoping type checker and other benefits.
```

Comment:
> practice since people are unlikely to accidentally use matchers and other Jest globals i

My experience is quite different. Folks created various helpers that are then used in implementation files and that gets into prod bundle. This can be ofc handled by proper lint rules, but I tend to use TS to its maximum potential.

btw jest was just an example :)

Reply:
I still can't remember seeing issues in this repo with people using testing utilities in production code by accident, but thinking about it more, one issue we've definitely run into is type pollution from Node globals contributing to IE11 breaks. In v8 component projects we usually have `lib` set to `es5, dom` to prevent references to ES6+ stuff that breaks IE 11--but `@types/node` (which should only be used in tests) includes a bunch of that stuff.

So while having to add an extra config file for every single project still makes me sad on general principal (harder to maintain), I can see how it might be helpful in this case.